--- ../../drbd/drbd/drbd-headers/linux/drbd.h	2023-02-17 14:26:26.854467352 +0000
+++ drbd-headers/linux/drbd.h	2023-02-17 14:26:29.370420069 +0000
@@ -26,6 +26,13 @@
 #ifndef DRBD_H
 #define DRBD_H
 
! compat: move those to drbd_windows.h
+#pragma warning (disable : 4100 4146 4221 4457 4456 4459)
+#pragma warning (disable : 4005 4018 4101 4115 4121 4127 4131 4152 4189 4200 4201 4204 4212 4218 4242 4244 4245 4267 4307 4389 4702 4706)
+/* Code analysis throws this warnings: */
+#pragma warning (disable : 26451 28719 6011 6385 6386 26453 33010 6001 28182)
+/* Enable all warnings throws lots of those warnings: */
+#pragma warning(disable: 4061 4062 4255 4388 4668 4820 5032 4711 5045)
+
 #include <asm/types.h>
 
 #ifdef __KERNEL__
