--- ../../drbd/drbd/drbd-headers/linux/genl_magic_struct.h	2023-02-17 14:26:26.854467352 +0000
+++ drbd-headers/linux/genl_magic_struct.h	2023-02-17 14:26:29.370420069 +0000
@@ -15,6 +15,7 @@
 
 #include <linux/netlink.h>
 #include <linux/genetlink.h>
! compat: needed? I think this header should move to linux subdir in the compat layer
+#include <drbd_wingenl.h>
 #ifdef __KERNEL__
 #include <net/genetlink.h>
 #else
@@ -122,7 +123,7 @@
 #define __str_field_def(attr_nr, attr_flag, name, maxlen) \
 	__str_field(attr_nr, attr_flag, name, maxlen)
 
! manual: __VA_ARGS__ in macro
-#define GENL_op_init(args...)	args
+#define GENL_op_init(...)	__VA_ARGS__
 #define GENL_doit(handler)		\
 	.doit = handler,		\
 	.flags = GENL_ADMIN_PERM,
