diff --git a/drbd/drbd_actlog.c b/drbd/drbd_actlog.c
index 78d4567..4c57c73 100644
--- a/drbd/drbd_actlog.c
+++ b/drbd/drbd_actlog.c
@@ -69,7 +69,7 @@ struct __packed al_transaction_on_disk {
 	/* Some reserved bytes.  Expected usage is a 64bit counter of
 	 * sectors-written since device creation, and other data generation tag
 	 * supporting usage */
-	__be32	__reserved[4];
+	__be32	__reserved_win[4];
 
 	/* --- 36 byte used --- */
 
