diff --git a/drbd/drbd_sender.c b/drbd/drbd_sender.c
index af7e7bd..be0825c 100644
--- a/drbd/drbd_sender.c
+++ b/drbd/drbd_sender.c
@@ -1262,26 +1262,7 @@ int w_e_end_data_req(struct drbd_work *w, int cancel)
 
 static bool all_zero(struct drbd_peer_request *peer_req)
 {
-	struct page *page = peer_req->page_chain.head;
-	unsigned int len = peer_req->i.size;
-
-	page_chain_for_each(page) {
-		unsigned int l = min_t(unsigned int, len, PAGE_SIZE);
-		unsigned int i, words = l / sizeof(long);
-		unsigned long *d;
-
-		d = drbd_kmap_atomic(page, KM_USER1);
-		for (i = 0; i < words; i++) {
-			if (d[i]) {
-				drbd_kunmap_atomic(d, KM_USER1);
-				return false;
-			}
-		}
-		drbd_kunmap_atomic(d, KM_USER1);
-		len -= l;
-	}
-
-	return true;
+	return false;
 }
 
 /**
