diff --git i/drbd/drbd_bitmap.c w/drbd/drbd_bitmap.c
index 51cfdc8d..4a58ca68 100644
--- i/drbd/drbd_bitmap.c
+++ w/drbd/drbd_bitmap.c
@@ -23,6 +23,8 @@

 #include "drbd_int.h"
 #include "drbd_dax_pmem.h"
+#include "drbd_windows.h"
+#include <linux/jiffies.h>

 #ifndef BITS_PER_PAGE
 #define BITS_PER_PAGE          (((ULONG_PTR)1) << (PAGE_SHIFT + 3))
@@ -1523,7 +1529,7 @@ unsigned int drbd_bm_clear_bits(struct drbd_device *device, unsigned int bitmap_
  *  0 ... bit not set
  * -1 ... first out of bounds access, stop testing for bits!
  */
-int drbd_bm_test_bit(struct drbd_peer_device *peer_device, const unsigned long bitnr)
+int drbd_bm_test_bit(struct drbd_peer_device *peer_device, unsigned long bitnr)
 {
 	struct drbd_bitmap *bitmap = peer_device->device->bitmap;
 	unsigned long irq_flags;
