diff --git a/drbd/drbd_sender.c b/drbd/drbd_sender.c
index af7e7bd..b802d9f 100644
--- a/drbd/drbd_sender.c
+++ b/drbd/drbd_sender.c
@@ -324,6 +324,7 @@ BIO_ENDIO_TYPE drbd_request_endio BIO_ENDIO_ARGS(struct bio *bio, int error)
 
 void drbd_csum_pages(struct crypto_ahash *tfm, struct page *page, void *digest)
 {
+#if 0
 	AHASH_REQUEST_ON_STACK(req, tfm);
 	struct scatterlist sg;
 
@@ -343,10 +344,12 @@ void drbd_csum_pages(struct crypto_ahash *tfm, struct page *page, void *digest)
 	ahash_request_set_crypt(req, NULL, digest, 0);
 	crypto_ahash_final(req);
 	ahash_request_zero(req);
+#endif
 }
 
 void drbd_csum_bio(struct crypto_ahash *tfm, struct bio *bio, void *digest)
 {
+#if 0
 	DRBD_BIO_VEC_TYPE bvec;
 	DRBD_ITER_TYPE iter;
 	AHASH_REQUEST_ON_STACK(req, tfm);
@@ -370,6 +373,7 @@ void drbd_csum_bio(struct crypto_ahash *tfm, struct bio *bio, void *digest)
 	ahash_request_set_crypt(req, NULL, digest, 0);
 	crypto_ahash_final(req);
 	ahash_request_zero(req);
+#endif
 }
 
 /* MAYBE merge common code with w_e_end_ov_req */
