diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index e207d80..56fcab0 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -3827,7 +3827,7 @@ void drbd_put_connection(struct drbd_connection *connection)
 	kref_sub(&connection->kref, refs, drbd_destroy_connection);
 }
 
-static int __init drbd_init(void)
+int __init drbd_init(void)
 {
 	int err;
 
@@ -4086,7 +4086,9 @@ int __init drbd_init(void)
 	spin_lock_init(&retry.lock);
 	INIT_LIST_HEAD(&retry.writes);
 
+#if 0
 	drbd_debugfs_init();
+#endif
 
 	pr_info("initialized. "
 	       "Version: " REL_VERSION " (api:%d/proto:%d-%d)\n",
