From db6e8c2780c2044617c00913846a22fec91ec454 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 6 Mar 2020 18:22:16 +0100
Subject: [PATCH 4/7] drbd_main: set_resource_options is a NOOP.

---
 drbd/drbd_main.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index 95672d4..85fdf53 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -3308,5 +3308,6 @@ static void wake_all_device_misc(struct drbd_resource *resource)
 
 int set_resource_options(struct drbd_resource *resource, struct res_opts *res_opts)
 {
+#if 0
 	struct drbd_connection *connection;
 	cpumask_var_t new_cpu_mask;
@@ -3388,7 +3389,9 @@ int set_resource_options(struct drbd_resource *resource, struct res_opts *res_op
 fail:
 	free_cpumask_var(new_cpu_mask);
 	return err;
-
+#else
+	return 0;
+#endif
 }
 
 struct drbd_resource *drbd_create_resource(const char *name,
-- 
2.17.0

