diff --git i/drbd/drbd_req.h w/drbd/drbd_req.h
index 7f546aec..948e853c 100644
--- i/drbd/drbd_req.h
+++ w/drbd/drbd_req.h
@@ -315,6 +315,7 @@ extern int __req_mod(struct drbd_request *req, enum drbd_req_event what,
 extern void complete_master_bio(struct drbd_device *device,
 		struct bio_and_error *m);
 extern void request_timer_fn(unsigned long data);
+static inline int request_timer_fn_win(KDPC *dpc, void * data, void *sysarg1, void *sysarg2) { (void)dpc; (void)sysarg1; (void)sysarg2; request_timer_fn((ULONG_PTR)data); return STATUS_SUCCESS; }
 extern void tl_restart(struct drbd_connection *connection, enum drbd_req_event what);
 extern void _tl_restart(struct drbd_connection *connection, enum drbd_req_event what);
 extern void drbd_queue_peer_ack(struct drbd_resource *resource, struct drbd_request *req);
