diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index f332da2..7ef6dc4 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -2328,9 +2328,11 @@ int drbd_send_dblock(struct drbd_peer_device *peer_device, struct drbd_request *
 
 	err = 0;
 	if (wsame) {
+#if 0
 		additional_size_command(peer_device->connection, DATA_STREAM,
 					bio_iovec(req->master_bio).bv_len);
 		err = __send_command(peer_device->connection, device->vnr, P_WSAME, DATA_STREAM);
+#endif
 	} else {
 		additional_size_command(peer_device->connection, DATA_STREAM, req->i.size);
 		err = __send_command(peer_device->connection, device->vnr, P_DATA, DATA_STREAM);
