diff -ru converted-sources/drbd/drbd_int.h converted-sources-drbd-thread-setup-nonstatic/drbd/drbd_int.h
--- converted-sources/drbd/drbd_int.h	2017-09-08 19:16:53.633020424 +0200
+++ converted-sources-drbd-thread-setup-nonstatic/drbd/drbd_int.h	2017-09-08 19:23:01.512868424 +0200
@@ -2341,6 +2341,7 @@
 extern int conn_send_state_req(struct drbd_connection *, int vnr, enum drbd_packet, union drbd_state, union drbd_state);
 extern int conn_send_twopc_request(struct drbd_connection *, int vnr, enum drbd_packet, struct p_twopc_request *);
 extern int drbd_send_peer_ack(struct drbd_connection *, struct drbd_request *);
+extern int drbd_thread_setup(void *arg);
 
 static inline void drbd_thread_stop(struct drbd_thread *thi)
 {
