From eec3d369467e84db29fe36b016ee40c31f18e954 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 25 Sep 2017 18:59:33 +0200
Subject: [PATCH 4/4] Do not emulate biosets.

---
 drbd/drbd_main.c       |   2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index 2523b0b..8274e6a 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -2706,12 +2706,12 @@ static int drbd_create_mempools(void)
 	drbd_io_bio_set = bioset_create(BIO_POOL_SIZE, 0, 0);
 	if (drbd_io_bio_set == NULL)
 		goto Enomem;
-#endif
 
 	drbd_md_io_bio_set = bioset_create(DRBD_MIN_POOL_PAGES, 0,
 					   BIOSET_NEED_BVECS);
 	if (drbd_md_io_bio_set == NULL)
 		goto Enomem;
+#endif
 
 	drbd_md_io_page_pool = mempool_create_page_pool(DRBD_MIN_POOL_PAGES, 0, '05WD');
 	if (drbd_md_io_page_pool == NULL)
