diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index 58b67a1..8240781 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -2121,7 +2121,7 @@ static int _drbd_send_bio(struct drbd_peer_device *peer_device, struct bio *bio)
 	alloc_send_buffer(connection, PAGE_SIZE, DATA_STREAM);
 	connection->send_buffer[DATA_STREAM].allocated_size = 0;
 
-	page.addr = bio->bio_databuf;
+        page.addr = bio->bi_io_vec[0].bv_page->addr;
 	err = _drbd_no_send_page(peer_device, &page, 0, bio->bi_size, 0);
 	if (err)
 		return err;
