diff --git a/drbd/drbd_bitmap.c b/drbd/drbd_bitmap.c
index b675c6f..cbdbb0c 100644
--- a/drbd/drbd_bitmap.c
+++ b/drbd/drbd_bitmap.c
@@ -1045,6 +1045,7 @@ static void drbd_bm_endio BIO_ENDIO_ARGS(struct bio *bio, blk_status_t status)
 	struct drbd_device *device = ctx->device;
 	struct drbd_bitmap *b = device->bitmap;
 	unsigned int idx = bm_page_to_idx(bio->bi_io_vec[0].bv_page);
+	struct page *page = NULL;
 
 	BIO_ENDIO_FN_START;
