From a27e463bb03ec4f4273e8111e31ba9c617ffd3ae Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Wed, 23 May 2018 15:50:13 +0200
Subject: [PATCH] Also kref_debug_put on failure in drbd_create_device().

---
 drbd/drbd_main.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index 2648056..64544b2 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -3753,12 +3753,14 @@ out_remove_peer_device:
 		kfree(peer_device);
 		kref_debug_put(&connection->kref_debug, 3);
 		kref_put(&connection->kref, drbd_destroy_connection);
+                kref_debug_put(&device->kref_debug, 1);
 	}
-
 	idr_remove(&resource->devices, vnr);
+        kref_debug_put(&device->kref_debug, 1);
 
 out_idr_remove_minor:
 	idr_remove(&drbd_devices, minor);
+        kref_debug_put(&device->kref_debug, 1);
 out_no_minor_idr:
 	if (locked)
 		spin_unlock_irq(&resource->req_lock);
@@ -3779,6 +3781,9 @@ out_no_disk:
 	blk_cleanup_queue(q);
 out_no_q:
 	kref_put(&resource->kref, drbd_destroy_resource);
+	kref_debug_put(&resource->kref_debug, 4);
+		/* kref debugging wants an extra put, see has_refs() */
+	kref_debug_put(&device->kref_debug, 4);
 	kref_debug_destroy(&device->kref_debug);
 	kfree(device);
 out_no_device:
-- 
2.7.4

