From ce07a1e86dcd21e395489b15786cc9634219c895 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 12 Oct 2018 19:07:12 +0200
Subject: [PATCH 1/5] New old patch for bitmap (was in 110-drbd_bitmap_C.patch)

---
 drbd/drbd_bitmap.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drbd/drbd_bitmap.c b/drbd/drbd_bitmap.c
index 9d84039..62efe7e 100644
--- a/drbd/drbd_bitmap.c
+++ b/drbd/drbd_bitmap.c
@@ -1121,7 +1121,11 @@ static void bm_page_io_async(struct drbd_bm_aio_ctx *ctx, int page_nr)
 	if (ctx->flags & BM_AIO_COPY_PAGES) {
 		page = mempool_alloc(&drbd_md_io_page_pool,
 				GFP_NOIO | __GFP_HIGHMEM);
-		copy_highpage(page, b->bm_pages[page_nr]);
+		expect(device, page);
+
+		page->private = b->bm_pages[page_nr]->private;
+		memcpy(page->addr, b->bm_pages[page_nr]->addr, PAGE_SIZE);
+
 		bm_store_page_idx(page, page_nr);
 	} else
 		page = b->bm_pages[page_nr];
-- 
2.17.0

