From 11e646185a99cd61e5bea2ec4ff8b0628624301b Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 6 Mar 2020 18:33:57 +0100
Subject: [PATCH 7/7] drbd_main: create block_device (was 557)

---
 drbd/drbd_main.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index bf9aeaa..65c2bc4 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -3658,15 +3658,21 @@ enum drbd_ret_code drbd_create_device(struct drbd_config_context *adm_ctx, unsig
 	int vnr = adm_ctx->volume;
 	enum drbd_ret_code err = ERR_NOMEM;
 	bool locked = false;
+	struct block_device *block_device;
 
 	device = minor_to_device(minor);
 	if (device)
 		return ERR_MINOR_OR_VOLUME_EXISTS;
 
+	block_device = bdget(MKDEV(DRBD_MAJOR, minor));
+	if (block_device == NULL)
+		return ERR_NO_DISK;
+
 	/* GFP_KERNEL, we are outside of all write-out paths */
 	device = kzalloc(sizeof(struct drbd_device), GFP_KERNEL, '0AWD');
 	if (!device)
-		return ERR_NOMEM;
+		goto out_no_device;
+
 	kref_init(&device->kref);
 	kref_debug_init(&device->kref_debug, &device->kref, &kref_class_device);
 
-- 
2.17.0

