From c2e33d121cf78f86da931fb8366bfced4c7b73b4 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 2 Nov 2018 15:05:58 +0100
Subject: [PATCH 1/2] Do not call generic_end_io_acct (was 509 patch)

---
 drbd/drbd_req.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drbd/drbd_req.c b/drbd/drbd_req.c
index 0bfbfe2..1652851 100644
--- a/drbd/drbd_req.c
+++ b/drbd/drbd_req.c
@@ -37,16 +37,20 @@ static bool drbd_may_do_local_read(struct drbd_device *device, sector_t sector,
 /* Update disk stats at start of I/O request */
 static void _drbd_start_io_acct(struct drbd_device *device, struct drbd_request *req)
 {
+#if 0
 	generic_start_io_acct(device->rq_queue, bio_data_dir(req->master_bio), req->i.size >> 9,
 			      &device->vdisk->part0);
+#endif
 }
 
 /* Update disk stats when completing request upwards */
 static void _drbd_end_io_acct(struct drbd_device *device, struct drbd_request *req)
 {
+#if 0
 	generic_end_io_acct(device->rq_queue, bio_data_dir(req->master_bio),
 			    &device->vdisk->part0,
 			    req->start_jif);
+#endif
 }
 #else
 static void _drbd_start_io_acct(struct drbd_device *device, struct drbd_request *req)
-- 
2.17.0

