From d4c6a34c9c17b7b10739bd4eb67be22719403dfa Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 18 Feb 2019 20:29:43 +0100
Subject: [PATCH 3/8] spin_lock_irq -> spin_lock_irqsave for drbd_req.c

---
 drbd/drbd_req.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drbd/drbd_req.c b/drbd/drbd_req.c
index 1652851..68854d8 100644
--- a/drbd/drbd_req.c
+++ b/drbd/drbd_req.c
@@ -2446,6 +2446,7 @@ static bool net_timeout_reached(struct drbd_request *net_req,
 void request_timer_fn(DRBD_TIMER_FN_ARG)
 {
 	KIRQL rcu_flags;
+	KIRQL flags;
 	struct drbd_device *device = DRBD_TIMER_ARG2OBJ(device, request_timer);
 	struct drbd_connection *connection;
 	struct drbd_request *req_read, *req_write;
@@ -2464,7 +2465,7 @@ void request_timer_fn(DRBD_TIMER_FN_ARG)
 	rcu_read_unlock(rcu_flags);
 
 	/* FIXME right now, this basically does a full transfer log walk *every time* */
-	spin_lock_irq(&device->resource->req_lock);
+	spin_lock_irqsave(&device->resource->req_lock, flags);
 	if (dt) {
 		ULONG_PTR write_pre_submit_jif, read_pre_submit_jif;
 		req_read = list_first_entry_or_null(&device->pending_completion[0], struct drbd_request, req_pending_local);
@@ -2549,7 +2550,7 @@ void request_timer_fn(DRBD_TIMER_FN_ARG)
 			end_state_change_locked(device->resource);
 		}
 	}
-	spin_unlock_irq(&device->resource->req_lock);
+	spin_unlock_irqrestore(&device->resource->req_lock, flags);
 
 	if (restart_timer) {
 		next_trigger_time = time_min_in_future(now, next_trigger_time, now + et);
-- 
2.17.0

