From b815c1895b46b027cfb626d37bec4a55b62f9e7e Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 18 Feb 2019 21:11:52 +0100
Subject: [PATCH 7/8] More spin_lock_irqsave for drbd_main.c

---
 drbd/drbd_main.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index 14bf423..f073d14 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -3000,14 +3000,15 @@ void drbd_reclaim_resource(struct rcu_head *rp)
 	struct drbd_resource *resource = container_of(rp, struct drbd_resource, rcu);
 	struct queued_twopc *q, *q1;
 	struct drbd_connection *connection, *tmp;
+	int flags;
 
-	spin_lock_irq(&resource->queued_twopc_lock);
+	spin_lock_irqsave(&resource->queued_twopc_lock, flags);
 	list_for_each_entry_safe(struct queued_twopc, q, q1, &resource->queued_twopc, w.list) {
 		list_del(&q->w.list);
 		kref_put(&q->connection->kref, drbd_destroy_connection);
 		kfree(q);
 	}
-	spin_unlock_irq(&resource->queued_twopc_lock);
+	spin_unlock_irqrestore(&resource->queued_twopc_lock, flags);
 
 	drbd_thread_stop_nowait(&resource->worker);
 
@@ -5175,11 +5176,12 @@ static int w_bitmap_io(struct drbd_work *w, int unused)
 void drbd_queue_pending_bitmap_work(struct drbd_device *device)
 {
 	ULONG_PTR flags;
+	int flags2;
 
 	spin_lock_irqsave(&device->pending_bitmap_work.q_lock, flags);
-	spin_lock(&device->resource->work.q_lock);
+	spin_lock_irqsave(&device->resource->work.q_lock, flags2);
 	list_splice_tail_init(&device->pending_bitmap_work.q, &device->resource->work.q);
-	spin_unlock(&device->resource->work.q_lock);
+	spin_unlock_irqrestore(&device->resource->work.q_lock, flags2);
 	spin_unlock_irqrestore(&device->pending_bitmap_work.q_lock, flags);
 	wake_up(&device->resource->work.q_wait);
 }
-- 
2.17.0

