From f54ec2e41dc8cbab29275f6444b5e1408d66e658 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 18 Feb 2019 21:19:28 +0100
Subject: [PATCH 8/8] More spin_lock_irqsave for drbd_sender.

---
 drbd/drbd_sender.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drbd/drbd_sender.c b/drbd/drbd_sender.c
index 14f2f10..784b96b 100644
--- a/drbd/drbd_sender.c
+++ b/drbd/drbd_sender.c
@@ -1915,6 +1915,7 @@ void drbd_start_resync(struct drbd_peer_device *peer_device, enum drbd_repl_stat
 	enum drbd_disk_state finished_resync_pdsk = D_UNKNOWN;
 	enum drbd_repl_state repl_state;
 	int r;
+	int flags;
 
 	spin_lock_irq(&device->resource->req_lock);
 	repl_state = peer_device->repl_state[NOW];
@@ -1996,11 +1997,11 @@ void drbd_start_resync(struct drbd_peer_device *peer_device, enum drbd_repl_stat
 		/* Forget potentially stale cached per resync extent bit-counts.
 		 * Open coded drbd_rs_cancel_all(device), we already have IRQs
 		 * disabled, and know the disk state is ok. */
-		spin_lock(&device->al_lock);
+		spin_lock_irqsave(&device->al_lock, flags);
 		lc_reset(peer_device->resync_lru);
 		peer_device->resync_locked = 0;
 		peer_device->resync_wenr = LC_FREE;
-		spin_unlock(&device->al_lock);
+		spin_unlock_irqrestore(&device->al_lock, flags);
 	}
 
 	unlock_all_resources();
-- 
2.17.0

