From 07f50f77ac8404c29d5e0b0edd445cbdad715a05 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Wed, 18 Mar 2020 11:11:10 +0100
Subject: [PATCH] drbd_sender: more types for from_timer() macro.

---
 drbd/drbd_sender.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drbd/drbd_sender.c b/drbd/drbd_sender.c
index 0ec1c3d..554165b 100644
--- a/drbd/drbd_sender.c
+++ b/drbd/drbd_sender.c
@@ -1899,7 +1899,7 @@ void drbd_rs_controller_reset(struct drbd_peer_device *peer_device)
 
 void start_resync_timer_fn(struct timer_list *t)
 {
-	struct drbd_peer_device *peer_device = from_timer(peer_device, t, start_resync_timer);
+	struct drbd_peer_device *peer_device = from_timer(peer_device, t, start_resync_timer, struct drbd_peer_device);
 	drbd_peer_device_post_work(peer_device, RS_START);
 }
 
@@ -2266,7 +2266,7 @@ static int do_md_sync(struct drbd_device *device)
 
 void repost_up_to_date_fn(struct timer_list *t)
 {
-	struct drbd_resource *resource = from_timer(resource, t, repost_up_to_date_timer);
+	struct drbd_resource *resource = from_timer(resource, t, repost_up_to_date_timer, struct drbd_resource);
 	drbd_post_work(resource, TRY_BECOME_UP_TO_DATE);
 }
 
-- 
2.17.0

