From 780fa3d672153e57f2d12876f82ef5ccbecbc235 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 8 May 2020 16:37:50 +0200
Subject: [PATCH 03/12] drbd_int.h: Always use KIRQL as datatype for IRQ flags.

---
 drbd/drbd_int.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drbd/drbd_int.h b/drbd/drbd_int.h
index 927d213..b1020d2 100644
--- a/drbd/drbd_int.h
+++ b/drbd/drbd_int.h
@@ -2248,7 +2248,7 @@ static inline void drbd_chk_io_error_(struct drbd_device *device,
 	int error, enum drbd_force_detach_flags forcedetach, const char *where)
 {
 	if (error) {
-		ULONG_PTR flags;
+		KIRQL flags;
 		spin_lock_irqsave(&device->resource->req_lock, flags);
 		__drbd_chk_io_error_(device, forcedetach, where);
 		spin_unlock_irqrestore(&device->resource->req_lock, flags);
@@ -2327,7 +2327,7 @@ void drbd_queue_work(struct drbd_work_queue *, struct drbd_work *);
 static inline void
 drbd_queue_work_if_unqueued(struct drbd_work_queue *q, struct drbd_work *w)
 {
-	ULONG_PTR flags;
+	KIRQL flags;
 	spin_lock_irqsave(&q->q_lock, flags);
 	if (list_empty_careful(&w->list))
 		list_add_tail(&w->list, &q->q);
-- 
2.17.0

