From 420c3552528428b8f309572cda89e26e5da00265 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Wed, 24 Mar 2021 17:38:40 +0100
Subject: [PATCH] drbd_req: for now no bi_disk field in struct bio.

---
 drbd/drbd_req.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drbd/drbd_req.c b/drbd/drbd_req.c
index ab4fad9..8c1ad97 100644
--- a/drbd/drbd_req.c
+++ b/drbd/drbd_req.c
@@ -2214,7 +2214,7 @@ static bool drbd_fail_request_early(struct drbd_device *device, struct bio *bio)
 
 blk_qc_t drbd_submit_bio(struct bio *bio)
 {
-	struct request_queue *q = bio->bi_disk->queue;
+	struct request_queue *q = bio->bi_bdev->bd_disk->queue;
 	struct drbd_device *device = (struct drbd_device *) q->queuedata;
 #ifdef CONFIG_DRBD_TIMING_STATS
 	ktime_t start_kt;
-- 
2.17.0

