From b42562f80ed7e3e8df53215aa32b395aae67a871 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 4 Oct 2021 16:37:16 +0200
Subject: [PATCH] drbd_sender: Use KIRQL for IRQ flags also in new code.

---
 drbd/drbd_sender.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drbd/drbd_sender.c b/drbd/drbd_sender.c
index 13e5f3c..8732ff6 100644
--- a/drbd/drbd_sender.c
+++ b/drbd/drbd_sender.c
@@ -467,7 +467,7 @@ int w_resync_timer(struct drbd_work *w, int cancel)
 	default:
 		if (atomic_read(&peer_device->rs_sect_in) >= peer_device->rs_in_flight) {
 			struct drbd_resource *resource = peer_device->device->resource;
-			ULONG_PTR irq_flags;
+			KIRQL irq_flags;
 			begin_state_change(resource, &irq_flags, 0);
 			peer_device->resync_active[NEW] = false;
 			end_state_change(resource, &irq_flags);
-- 
2.17.0

