From 331a070cc768209db94e2b0aff4e5dd8d6b866fd Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 28 Mar 2022 12:00:53 -0700
Subject: [PATCH 09/12] drbd_req: pass code analysis and codeql tests.

---
 drbd/drbd_req.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drbd/drbd_req.c b/drbd/drbd_req.c
index 92c73eb..99fcb94 100644
--- a/drbd/drbd_req.c
+++ b/drbd/drbd_req.c
@@ -2406,7 +2406,7 @@ void request_timer_fn(struct timer_list *t)
 			if (connection->cstate[NOW] == C_CONNECTED) {
 				ko_count = nc->ko_count;
 				timeout = nc->timeout;
-				ent = timeout * HZ/10 * ko_count;
+				ent = ((ULONG_PTR)timeout) * HZ/10 * ko_count;
 			}
 		}
 		rcu_read_unlock(rcu_flags);
-- 
2.35.1

