From 14485d3b84133e1333a9673f7ffe5667c43996fd Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 31 Mar 2025 13:28:33 +0000
Subject: [PATCH 2/2] ioctl for testing if the bus device works.

---
 drbd/drbd-headers/windrbd/windrbd_ioctl.h | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drbd/drbd-headers/windrbd/windrbd_ioctl.h b/drbd/drbd-headers/windrbd/windrbd_ioctl.h
index 6243537..80142ad 100644
--- a/drbd/drbd-headers/windrbd/windrbd_ioctl.h
+++ b/drbd/drbd-headers/windrbd/windrbd_ioctl.h
@@ -366,4 +366,17 @@ struct windrbd_minor_mount_point {
 
 #define IOCTL_WINDRBD_ROOT_DRBD_OP_IS_KNOWN CTL_CODE(WINDRBD_ROOT_DEVICE_TYPE, 23, METHOD_BUFFERED, FILE_ANY_ACCESS)
 
+/* Check if the WinDRBD virtual bus device is working correctly
+ *
+ * Input: none
+ * Output: 0 - does NOT work correctly 1 - works as expected
+ *
+ * Sometimes the WinDRBD bus device does not work correctly after
+ * a new install. The reason for that is yet unknown but we know
+ * how to fix it. In order to allow the installer to apply the fix
+ * we need a possibility to check if the bus device works as expected.
+ */
+
+#define IOCTL_WINDRBD_ROOT_BUS_DEVICE_IS_WORKING CTL_CODE(WINDRBD_ROOT_DEVICE_TYPE, 24, METHOD_BUFFERED, FILE_ANY_ACCESS)
+
 #endif
-- 
2.17.1

