From de93438cba5063177950af7c321425c5b63660e2 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 6 Oct 2017 20:31:32 +0200
Subject: [PATCH 12/15] printk in drbd_destroy_device()

---
 drbd/drbd_main.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drbd/drbd_main.c b/drbd/drbd_main.c
index 846e12c..2a5763d 100644
--- a/drbd/drbd_main.c
+++ b/drbd/drbd_main.c
@@ -2804,6 +2804,8 @@ void drbd_destroy_device(struct kref *kref)
 	struct drbd_device *device = container_of(kref, struct drbd_device, kref);
 	struct drbd_peer_device *peer_device, *tmp;
 
+printk(KERN_INFO "drbd_destroy_device %p\n", device);
+
 	/* cleanup stuff that may have been allocated during
 	 * device (re-)configuration or state changes */
 
-- 
2.8.0-rc4

